opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java

```diff
@@ -0,0 +1,112 @@
+ *
+ * @author Pavol Loffay
+ */
+public class TracingInterceptor implements Interceptor {
+
+    private Tracer tracer;
+    private List<SpanDecorator> decorators;
+
+    private ThreadLocal<Connection> activeConnection = new ThreadLocal<>();
+
+    /**
+     * Create tracing interceptor. Interceptor has to be added to {@link OkHttpClient.Builder#addInterceptor(Interceptor)}
+     * and {@link OkHttpClient.Builder#addNetworkInterceptor(Interceptor)}.
+     *
+     * @param tracer tracer
+     * @param decorators decorators
+     */
+    public TracingInterceptor(Tracer tracer, List<SpanDecorator> decorators) {
+        this.tracer = tracer;
+        this.decorators = new ArrayList<>(decorators);
+    }
+
+    /**
+     * Add tracing interceptors to client builder.
+     *
+     * @param okBuilder client builder
+     * @param tracer tracer
+     * @param decorators span decorators
+     * @return client builder with added tracing interceptor
+     */
+    public static OkHttpClient.Builder addTracing(OkHttpClient.Builder okBuilder,
+                                                  Tracer tracer, List<SpanDecorator> decorators) {
+
+        TracingInterceptor tracingInterceptor = new TracingInterceptor(tracer, decorators);
+        return okBuilder.addInterceptor(tracingInterceptor)
+                    .addNetworkInterceptor(tracingInterceptor);
+    }
+
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+        Response response;
+
+        // application interceptor?
+        if (chain.connection() == null) {
+            Tracer.SpanBuilder spanBuilder = tracer.buildSpan(chain.request().method())
+                    .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT);
+
+            Object tag = chain.request().tag();
+            if (tag != null && tag instanceof SpanContext) {
+                spanBuilder.asChildOf((SpanContext) tag);
+            }
+
+            Span span = spanBuilder.start();
+
+            for (SpanDecorator spanDecorator: decorators) {
+                spanDecorator.onRequest(chain.request(), span);
+            }
+
+            Request.Builder requestBuilder = chain.request().newBuilder();
+            tracer.inject(span.context(), Format.Builtin.HTTP_HEADERS, new RequestBuilderInjectAdapter(requestBuilder));
+
+            try {
+                response = chain.proceed(requestBuilder.build());
+                for (SpanDecorator spanDecorator: decorators) {
+                    spanDecorator.onResponse(response, activeConnection.get(), span);
+                }
+            } catch (Throwable ex) {
+                for (SpanDecorator spanDecorator: decorators) {
+                    spanDecorator.onError(ex, span);
+                }
+                throw ex;
+            } finally {
+                span.finish();
+            }
+        } else {
+            response = chain.proceed(chain.request());
+            activeConnection.set(chain.connection());
```

Comment:
There's no guarantee that application and network interceptors run on the same thread for the same call.

That said, that's how it works today and there's no particular plans to change that.